File: src/main/java/dev/dnpm/etl/processor/services/KeycloakTokenService.java
package dev.dnpm.etl.processor.services;



import org.springframework.beans.factory.annotation.Value;
import org.springframework.http.*;
import org.springframework.stereotype.Service;
import org.springframework.web.client.RestTemplate;
import java.util.Map;

@Service
public class KeycloakTokenService {

    @Value("${keycloak.token-url}")
    private String tokenUrl;

    @Value("${keycloak.client-id}")
    private String clientId;

    @Value("${keycloak.client-secret}")
    private String clientSecret;

    @Value("${keycloak.username}")
    private String username;

    @Value("${keycloak.password}")
    private String password;

    private String cachedToken;

    public String getAccessToken() {
        if (cachedToken != null) return cachedToken;

        RestTemplate restTemplate = new RestTemplate();

        HttpHeaders headers = new HttpHeaders();
        headers.setContentType(MediaType.APPLICATION_FORM_URLENCODED);

        String body = String.format(
                "client_id=%s&client_secret=%s&username=%s&password=%s&grant_type=password",
                clientId, clientSecret, username, password
        );

        HttpEntity<String> entity = new HttpEntity<>(body, headers);

        ResponseEntity<Map> response = restTemplate.postForEntity(tokenUrl, entity, Map.class);

        if (response.getStatusCode().is2xxSuccessful() && response.getBody() != null) {
            cachedToken = (String) response.getBody().get("access_token");
            return cachedToken;
        }
        throw new RuntimeException("Failed to get Keycloak token: " + response.getStatusCode());
    }
}

File: src/main/kotlin/dev/dnpm/etl/processor/services/KeycloakTokenService.kt
package dev.dnpm.etl.processor.services

import dev.dnpm.etl.processor.config.KeycloakProperties
import org.springframework.web.client.RestTemplate
import org.springframework.http.HttpEntity
import org.springframework.http.HttpHeaders
import org.springframework.http.MediaType
import org.springframework.stereotype.Service
import java.util.Base64

@Service
class KeycloakTokenService(
    private val restTemplate: RestTemplate,
    private val keycloakProperties: KeycloakProperties
) {

    fun getToken(client: String): String {
        val props = when (client.lowercase()) {
            "gpas" -> keycloakProperties.gpas
            "diz" -> keycloakProperties.diz
            else -> throw IllegalArgumentException("Unknown Keycloak client: $client")
        }

        val headers = HttpHeaders()
        headers.contentType = MediaType.APPLICATION_FORM_URLENCODED

        val body = mapOf(
            "client_id" to props.clientId,
            "client_secret" to props.clientSecret,
            "username" to props.username,
            "password" to props.password,
            "grant_type" to "password"
        ).map { "${it.key}=${it.value}" }.joinToString("&")

        val entity = HttpEntity(body, headers)
        val response = restTemplate.postForObject(props.tokenUrl, entity, Map::class.java)

        @Suppress("UNCHECKED_CAST")
        return response?.get("access_token") as? String
            ?: throw IllegalStateException("Failed to get Keycloak token for client ${props.clientId}")
    }
}
